src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,92 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int INITIALIZATION_SIZE = 10;
+    private static final double ARRAY_RATIO = 1.5;
+    private T[] values;
+    private int size;
 
+    public ArrayList() {
+        values = (T[]) new Object[INITIALIZATION_SIZE];
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        growArrayIfFull();
+        values[size] = value;
+        size++;
+    }
 
+    public void add(T value, int index) {
+        checkIndex(index, size);
+        T[] newValues = Arrays.copyOf(values, values.length + 1);
+        System.arraycopy(newValues, index, newValues, index + 1, size - index);
+        newValues[index] = value;
+        values = newValues;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index, size - 1);
+        values[index] = value;
     }
 
-    @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        T removedElement = values[index];
+        System.arraycopy(values, index + 1, values, index,size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < values.length; i++) {
+            if (element == values[i] || element != null && element.equals(values[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growArrayIfFull() {
+        if (values.length == size) {
+            System.arraycopy(values, 0,
+                    values = (T[]) new Object[(int) (size * ARRAY_RATIO)], 0, size);
```

Comment:
Consider to do like that, because write many logic things in one line is not a good approach.
```suggestion
int newCapacity = (int) (size * ARRAY_RATIO);
T[] newElements = (T[]) new Object[newCapacity];
System.arraycopy(elements, 0, newElements, 0, size);
values = newArray;
```